openwisp_firmware_upgrader/base/models.py

```diff
@@ -626,6 +630,17 @@ def upgrade(self, recoverable=True):
             self.device.devicefirmware.installed = True
             self.device.devicefirmware.save(upgrade=False)
 
+    def clean(self):
+        super().clean()
+        if self.upgrade_options:
+            # Only store the keys that were set
+            pop_keys = []
+            for key in self.upgrade_options.keys():
+                if not self.upgrade_options[key]:
+                    pop_keys.append(key)
+            for key in pop_keys:
+                self.upgrade_options.pop(key, None)
+
```

Comment:
@nemesisdesign I want to replace this with proper JSONSchema validation. I am planning to achieve this with the following code 

```python
    def clean(self):
        super().clean()
        self.validate_upgrade_options(self)

    def validate_upgrade_options(self):
        DeviceConnection = swapper.load_model('connection', 'DeviceConnection')
        try:
            conn = DeviceConnection.get_working_connection(self.device)
        except NoWorkingDeviceConnectionError:
            raise ValidationError(_('This device does not have a working connection.'))
        try:
            upgrader_class = app_settings.UPGRADERS_MAP[conn.update_strategy]
            upgrader_class = import_string(upgrader_class)
        except (AttributeError, ImportError):
            raise ValidationError(_('This device does not have a supported upgrader.'))
        try:
            jsonschema.Draft4Validator(upgrader_class.SCHEMA).validate(self.upgrade_options)
        except jsonschema.SchemaError as error:
            raise ValidationError({'upgrade_options': error.message})
```

But, this will make validation very slow and prone to errors. Shall we do schema validation directly at upgrader class?

Reply:
I think the simplest solution to the problem of automatically figuring out which upgrader class to use is the following:

- query the DB for the available device connection, filter only SSH connections
- get the first available SSH connection
  (it doesn't matter if it works or not, because there cannot be connections for two different firmware types but rather two different credentials, eg: one with password and one with SSH key, or different SSH keys, which doesn't make any difference for determining the firmware type)
  this means to not hit the network to test if the connection works or not
- now from the connection you can figure out what upgrader to use `openwisp_firmware_upgrader.settings.UPGRADERS_MAP`

This will make the system extensible because when custom upgrader classes are used are set via the `UPGRADERS_MAP` setting, that is, if I change the default firmware upgrader related to the OpenWrt ssh connection, and the custom upgrader class has a different schema, this schema shall be picked automatically by the system. That's the whole point of this effort.